File: src/test/java/com/fwdekker/randomness/decimal/DecimalSettingsDialogTest.java
package com.fwdekker.randomness.decimal;

import com.intellij.openapi.ui.ValidationInfo;
import org.assertj.swing.edt.GuiActionRunner;
import org.assertj.swing.fixture.FrameFixture;
import org.assertj.swing.junit.testcase.AssertJSwingJUnitTestCase;
import org.junit.Test;

import static org.assertj.core.api.Assertions.assertThat;
import static org.assertj.swing.fixture.Containers.showInFrame;


/**
 * GUI tests for {@link DecimalSettingsDialog}.
 */
public final class DecimalSettingsDialogTest extends AssertJSwingJUnitTestCase {
    private static final double DEFAULT_MIN_VALUE = 157.61;
    private static final double DEFAULT_MAX_VALUE = 408.68;
    private static final int DEFAULT_DECIMAL_COUNT = 5;

    private DecimalSettings decimalSettings;
    private DecimalSettingsDialog decimalSettingsDialog;
    private FrameFixture frame;


    @Override
    protected void onSetUp() {
        decimalSettings = new DecimalSettings();
        decimalSettings.setMinValue(DEFAULT_MIN_VALUE);
        decimalSettings.setMaxValue(DEFAULT_MAX_VALUE);
        decimalSettings.setDecimalCount(DEFAULT_DECIMAL_COUNT);

        decimalSettingsDialog = GuiActionRunner.execute(() -> new DecimalSettingsDialog(decimalSettings));
        frame = showInFrame(robot(), decimalSettingsDialog.getContentPane());
    }


    @Test
    public void testDefaultMinValue() {
        frame.spinner("minValue").requireValue(DEFAULT_MIN_VALUE);
    }

    @Test
    public void testDefaultMaxValue() {
        frame.spinner("maxValue").requireValue(DEFAULT_MAX_VALUE);
    }

    @Test
    public void testDefaultDecimalCount() {
        frame.spinner("decimalCount").requireValue(DEFAULT_DECIMAL_COUNT);
    }

    @Test
    public void testDefaultIsValid() {
        assertThat(decimalSettingsDialog.doValidate()).isNull();
    }

    @Test
    public void testMinValueType() {
        GuiActionRunner.execute(() -> frame.spinner("minValue").target().setValue(396.28f));

        final ValidationInfo validationInfo = decimalSettingsDialog.doValidate();

        assertThat(validationInfo).isNotNull();
        assertThat(validationInfo.component).isEqualTo(frame.spinner("minValue").target());
        assertThat(validationInfo.message).isEqualTo("Minimum value must be a decimal.");
    }

    @Test
    public void testMaxValueType() {
        GuiActionRunner.execute(() -> frame.spinner("maxValue").target().setValue(355.28f));

        final ValidationInfo validationInfo = decimalSettingsDialog.doValidate();

        assertThat(validationInfo).isNotNull();
        assertThat(validationInfo.component).isEqualTo(frame.spinner("maxValue").target());
        assertThat(validationInfo.message).isEqualTo("Maximum value must be a decimal.");
    }

    @Test
    public void testDecimalCountType() {
        GuiActionRunner.execute(() -> frame.spinner("decimalCount").target().setValue(166.98));

        final ValidationInfo validationInfo = decimalSettingsDialog.doValidate();

        assertThat(validationInfo).isNotNull();
        assertThat(validationInfo.component).isEqualTo(frame.spinner("decimalCount").target());
        assertThat(validationInfo.message).isEqualTo("Decimal count must be an integer.");
    }

    @Test
    public void testMaxValueGreaterThanMinValue() {
        frame.spinner("maxValue").enterTextAndCommit(Integer.toString((int) DEFAULT_MIN_VALUE - 1));

        final ValidationInfo validationInfo = decimalSettingsDialog.doValidate();

        assertThat(validationInfo).isNotNull();
        assertThat(validationInfo.component).isEqualTo(frame.spinner("maxValue").target());
        assertThat(validationInfo.message).isEqualTo("Maximum value cannot be smaller than minimum value.");
    }

    @Test
    public void testNegativeDecimalCount() {
        frame.spinner("decimalCount").enterTextAndCommit(Double.toString(-851));

        final ValidationInfo validationInfo = decimalSettingsDialog.doValidate();

        assertThat(validationInfo).isNotNull();
        assertThat(validationInfo.component).isEqualTo(frame.spinner("decimalCount").target());
        assertThat(validationInfo.message).isEqualTo("Decimal count must be at least 0.");
    }

    @Test
    public void testSaveSettings() {
        frame.spinner("minValue").enterText("418.63");
        frame.spinner("maxValue").enterText("858.59");
        frame.spinner("decimalCount").enterText("99");

        GuiActionRunner.execute(() -> decimalSettingsDialog.saveSettings());

        assertThat(decimalSettings.getMinValue()).isEqualTo(418.63);
        assertThat(decimalSettings.getMaxValue()).isEqualTo(858.59);
        assertThat(decimalSettings.getDecimalCount()).isEqualTo(99);
    }
}

File: src/test/kotlin/com/fwdekker/randomness/decimal/DecimalSettingsDialogTest.kt
package com.fwdekker.randomness.decimal

import org.assertj.core.api.Assertions.assertThat
import org.assertj.swing.edt.FailOnThreadViolationRepaintManager
import org.assertj.swing.edt.GuiActionRunner
import org.assertj.swing.fixture.Containers.showInFrame
import org.assertj.swing.fixture.FrameFixture
import org.jetbrains.spek.api.Spek
import org.jetbrains.spek.api.dsl.describe
import org.jetbrains.spek.api.dsl.it


/**
 * GUI tests for [DecimalSettingsDialog].
 */
object DecimalSettingsDialogTest : Spek({
    lateinit var decimalSettings: DecimalSettings
    lateinit var decimalSettingsDialog: DecimalSettingsDialog
    lateinit var frame: FrameFixture


    beforeGroup {
        FailOnThreadViolationRepaintManager.install()
    }

    beforeEachTest {
        decimalSettings = DecimalSettings()
        decimalSettings.minValue = 157.61
        decimalSettings.maxValue = 408.68
        decimalSettings.decimalCount = 5
        decimalSettings.showTrailingZeroes = false
        decimalSettings.groupingSeparator = "_"
        decimalSettings.decimalSeparator = "."

        decimalSettingsDialog =
            GuiActionRunner.execute<DecimalSettingsDialog> { DecimalSettingsDialog(decimalSettings) }
        frame = showInFrame(decimalSettingsDialog.createCenterPanel())
    }

    afterEachTest {
        frame.cleanUp()
    }


    describe("loading settings") {
        it("loads the settings' minimum value") {
            frame.spinner("minValue").requireValue(157.61)
        }

        it("loads the settings' maximum value") {
            frame.spinner("maxValue").requireValue(408.68)
        }

        it("loads the settings' decimal count") {
            frame.spinner("decimalCount").requireValue(5)
        }

        it("loads the settings' value for showing trailing zeroes") {
            frame.checkBox("showTrailingZeroes").requireSelected(false)
        }

        it("loads the settings' grouping separator") {
            frame.radioButton("groupingSeparatorNone").requireSelected(false)
            frame.radioButton("groupingSeparatorPeriod").requireSelected(false)
            frame.radioButton("groupingSeparatorComma").requireSelected(false)
            frame.radioButton("groupingSeparatorUnderscore").requireSelected(true)
        }

        it("loads the settings' decimal separator") {
            frame.radioButton("decimalSeparatorComma").requireSelected(false)
            frame.radioButton("decimalSeparatorPeriod").requireSelected(true)
        }
    }

    describe("input handling") {
        it("truncates decimals in the decimal count") {
            GuiActionRunner.execute { frame.spinner("decimalCount").target().value = 693.57f }

            frame.spinner("decimalCount").requireValue(693)
        }
    }

    describe("validation") {
        it("passes for the default settings") {
            GuiActionRunner.execute { decimalSettingsDialog.loadSettings(DecimalSettings()) }

            assertThat(decimalSettingsDialog.doValidate()).isNull()
        }

        describe("value range") {
            it("fails if the minimum value is greater than the maximum value") {
                GuiActionRunner.execute {
                    frame.spinner("minValue").target().value = 46.16
                    frame.spinner("maxValue").target().value = 45.16
                }

                val validationInfo = decimalSettingsDialog.doValidate()

                assertThat(validationInfo).isNotNull()
                assertThat(validationInfo?.component).isEqualTo(frame.spinner("maxValue").target())
                assertThat(validationInfo?.message).isEqualTo("The maximum should not be smaller than the minimum.")
            }

            it("fails if the range size overflows") {
                GuiActionRunner.execute {
                    frame.spinner("minValue").target().value = -1E53
                    frame.spinner("maxValue").target().value = 1E53
                }

                val validationInfo = decimalSettingsDialog.doValidate()

                assertThat(validationInfo).isNotNull()
                assertThat(validationInfo?.component).isEqualTo(frame.spinner("maxValue").target())
                assertThat(validationInfo?.message).isEqualTo("The range should not exceed 1.0E53.")
            }
        }

        describe("decimal count") {
            it("passes if the decimal count is zero") {
                GuiActionRunner.execute { frame.spinner("decimalCount").target().value = 0 }

                assertThat(decimalSettingsDialog.doValidate()).isNull()
            }

            it("fails if the decimal count is negative") {
                GuiActionRunner.execute { frame.spinner("decimalCount").target().value = -851 }

                val validationInfo = decimalSettingsDialog.doValidate()

                assertThat(validationInfo).isNotNull()
                assertThat(validationInfo?.component).isEqualTo(frame.spinner("decimalCount").target())
                assertThat(validationInfo?.message).isEqualTo("Please enter a value greater than or equal to 0.")
            }
        }
    }

    describe("saving settings") {
        it("correctly saves settings to a settings object") {
            GuiActionRunner.execute {
                frame.spinner("minValue").target().value = 112.54
                frame.spinner("maxValue").target().value = 644.74
                frame.spinner("decimalCount").target().value = 485
                frame.checkBox("showTrailingZeroes").target().isSelected = false
                frame.radioButton("groupingSeparatorUnderscore").target().isSelected = true
                frame.radioButton("decimalSeparatorComma").target().isSelected = true
            }

            decimalSettingsDialog.saveSettings()

            assertThat(decimalSettings.minValue).isEqualTo(112.54)
            assertThat(decimalSettings.maxValue).isEqualTo(644.74)
            assertThat(decimalSettings.decimalCount).isEqualTo(485)
            assertThat(decimalSettings.showTrailingZeroes).isEqualTo(false)
            assertThat(decimalSettings.groupingSeparator).isEqualTo("_")
            assertThat(decimalSettings.decimalSeparator).isEqualTo(",")
        }
    }
})
